File: python/software/firmware.py
from datetime import date
from interfaces.executable import Executable
from interfaces.upgradable import Upgradable

class Firmware(Executable, Upgradable):
    def __init__(self, firmware_version="", supported_hardware="", release_date=None, installed=False):
        self._firmware_version = firmware_version
        self._supported_hardware = supported_hardware
        self._release_date = release_date if release_date is not None else date.today()
        self._installed = installed

    # Setters
    def set_firmware_version(self, firmware_version):
        self._firmware_version = firmware_version

    def set_supported_hardware(self, supported_hardware):
        self._supported_hardware = supported_hardware

    def set_release_date(self, release_date):
        self._release_date = release_date

    def set_installed(self, installed):
        self._installed = installed

    # Getters
    def get_firmware_version(self):
        return self._firmware_version

    def get_supported_hardware(self):
        return self._supported_hardware

    def get_release_date(self):
        return self._release_date

    def is_installed(self):
        return self._installed

    def execute(self):
        print("Firmware executed")

    def terminate(self):
        print("Firmware terminated")

    def upgrade(self):
        print("Firmware upgraded")
        return True

    def downgrade(self):
        print("Firmware downgraded")
        return True

File: cpp/software/Firmware.cpp
#pragma once
#include <bits/stdc++.h>

#include "../utils/Date.cpp"
#include "../interfaces/Executable.cpp"
#include "../interfaces/Upgradable.cpp"

class Firmware : public Executable, public Upgradable
{
protected:
    std::string firmwareVersion;
    std::string supportedHardware;
    Date releaseDate;
    bool installed;

public:
    // Constructors
    Firmware() : firmwareVersion(""), supportedHardware(""), releaseDate(), installed(false) {}
    Firmware(const std::string &version, const std::string &hardware, const Date &date)
        : firmwareVersion(version), supportedHardware(hardware), releaseDate(date), installed(false) {}

    // Destructor
    ~Firmware() {}

    // Setters
    void setFirmwareVersion(const std::string &version)
    {
        firmwareVersion = version;
    }

    void setSupportedHardware(const std::string &hardware)
    {
        supportedHardware = hardware;
    }

    void setReleaseDate(const Date &date)
    {
        releaseDate = date;
    }

    void setInstalled(bool isInstalled)
    {
        installed = isInstalled;
    }

    // Getters
    std::string getFirmwareVersion() const
    {
        return firmwareVersion;
    }

    std::string getSupportedHardware() const
    {
        return supportedHardware;
    }

    Date getReleaseDate() const
    {
        return releaseDate;
    }

    bool isInstalled() const
    {
        return installed;
    }

    void execute() override
    {
        printf("Firmware executed\n");
    }

    void terminate() override
    {
        printf("Firmware terminated\n");
    }

    bool upgrade() override
    {
        printf("Firmware upgraded\n");
        return true;
    }

    bool downgrade() override
    {
        printf("Firmware downgraded\n");
        return true;
    }
};
